src/sov/mod.rs

```diff
@@ -0,0 +1,206 @@
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Allocation {
+    pub partner: String,
+    pub percentage: i64,
+}
+
+impl From<&mut Settings> for HandlerResult<SOVManager> {
+    fn from(settings: &mut Settings) -> Self {
+        let (source_url, last_response) = if settings.sov_source.starts_with("gs://") {
+            (
+                Some(
+                    settings
+                        .sov_source
+                        .parse::<url::Url>()
+                        .map_err(|e| {
+                            ConfigError::Message(format!(
+                                "Unable to parse SOV URL '{}': {:?}",
+                                &settings.sov_source, e
+                            ))
+                        })
+                        .unwrap(),
+                ),
+                None,
+            )
+        } else if Path::new(&settings.sov_source).exists() {
+            let response = match read_to_string(&settings.sov_source) {
+                Ok(contents) => serde_json::from_str(&contents)
+                    .map_err(|e| {
+                        ConfigError::Message(format!(
+                            "Could not parse SOV settings from file '{}': {:?}",
+                            settings.sov_source, e
+                        ))
+                    })
+                    .unwrap(),
+                Err(e) => panic!(
+                    "Could not read file '{}' with SOV settings: {:?}",
+                    &settings.sov_source, e
+                ),
+            };
+            (
+                None,
+                Some(LastResponse {
+                    response,
+                    updated: Utc::now(),
+                }),
+            )
+        } else if !settings.sov_source.is_empty() {
+            let response = serde_json::from_str::<SOVResponse>(&settings.sov_source)
+                .map_err(|e| ConfigError::Message(format!("Could not parse SOV settings: {:?}", e)))
+                .unwrap();
+            (
+                None,
+                Some(LastResponse {
+                    response,
+                    updated: Utc::now(),
+                }),
+            )
+        } else {
+            (None, None)
+        };
+
+        if last_response.as_ref().is_some() {
+            Ok(SOVManager {
+                refresh_rate: Duration::from_secs(settings.sov_refresh_rate_secs),
+                source_url,
+                encoded_sov: Some(base64::engine::general_purpose::STANDARD_NO_PAD.encode(
+                    serde_json::to_string(&last_response.as_ref().unwrap().response).unwrap(),
+                )),
+                last_response,
+            })
+        } else {
+            Ok(SOVManager {
+                refresh_rate: Duration::from_secs(settings.sov_refresh_rate_secs),
+                source_url,
+                encoded_sov: None,
+                last_response,
+            })
+        }
+    }
+}
```

Comment:
gut feeling tells me there's a more simple way to do this

Reply:
I think we can slightly simplify this as follows. Note that I used `HandlerError::Internal` here for parsing errors and defer error handling to upstream code instead of unwrapping.

```rs
    fn from(settings: &mut Settings) -> Self {
        // Try with the remote source first.
        if settings.sov_source.starts_with("gs://") {
            let Ok(source_url) = settings.sov_source.parse::<url::Url>() else {
                return Err(
                    HandlerErrorKind::Internal(
                        format!("Unable to parse SOV URL '{}'", &settings.sov_source)
                    ).into()
                );
            };

            return Ok(SOVManager {
                source_url: Some(source_url),
                refresh_rate: Duration::from_secs(settings.sov_refresh_rate_secs),
                encoded_sov: None,
                last_response: None,
            });
        }

        // Then check if it's a local settings file or an inline settings string.
        let sov = if Path::new(&settings.sov_source).exists() {
            let Ok(sov) =
                read_to_string(&settings.sov_source)
                    .map_err(|_| Err::<String, &str>("Unable to read SOV settings file"))
                    .and_then(|content| {
                        serde_json::from_str::<SOVResponse>(&content)
                            .map_err(|_| Err("Unable to load SOV settings from JSON"))
                    }) else {
                        return Err(
                            HandlerErrorKind::Internal(
                                format!("Unable to parse SOV settings from file '{}'", settings.sov_source)
                            ).into()
                        );
                    };
            sov
        } else {
            // Presume it's an inline SOV settings string.
            let Ok(sov) = serde_json::from_str::<SOVResponse>(&settings.sov_source) else {
                return Err(
                    HandlerErrorKind::Internal(
                        format!("Could not parse SOV settings inline: {:?}", &settings.sov_source)
                    ).into()
                );
            };
            sov
        };

        Ok(SOVManager {
            source_url: None,
            refresh_rate: Duration::from_secs(settings.sov_refresh_rate_secs),
            encoded_sov: Some(
                // Unwrapping is safe here since the parsing above ensures it's a valid `SOVResponse`.
                base64::engine::general_purpose::STANDARD_NO_PAD
                    .encode(serde_json::to_string(&sov).unwrap()),
            ),
            last_response: Some(LastResponse {
                response: sov,
                updated: Utc::now(),
            }),
        })
    }
```